### DISCUSSION TITLE
V2 and Props

### DISCUSSION THREAD

**Discussion by pauleveritt** [Finistere/antidote#68] [General]
I'm back in "book writing" mode. I'd like to explore "props" from the previous, so I can bring my antidom repo into the book.

*tl;dr Where do "props" get into injection, how are they used during instantiation?*

Imagine I have "template": `<p><{Greeting} name="Antidote" /></p>`. This results in: `<p>Hello Antidote!</p>`
`Hello` comes from a component local variable, `Antidote` from a passed-in "prop", and `!` is injected.

Maybe I have:

```python
@inject
def Greeting(
        punctuation: str = inject[get_punctuation],
        customer_name: str = [MAGIC HAPPENS HERE],
) -> str:
    salutation = "Hello"
    return f"{salutation} {customer_name}{punctuation}"
```

Maybe I have class-based:

```python
@injectable(scope="deterministic")
@dataclass
class Greeting:
    punctuation: str = inject[get_punctuation],
    customer_name: str = [MAGIC HAPPENS HERE],
    salutation: str = "Hello"

    def __call__(self) -> str:
        return f"{self.salutation} {self.customer_name}{self.punctuation}"
```

The second version offers the possibility to pass the props into the __call__. On one hand, that's unattractive, as results will never get cached. OTOH, the *instance of these props* would have to be in `state` somewhere. Which might be a good answer.

Or perhaps I need to write a provider. Maybe I need to anyway, to customize the `.update` of injection?

Last point: a prop should be able to override an injected value. In the above, I should be able to pass in my own punctuation for a specific usage:  `<p><{Greeting} name="Antidote" punctuation="??"  /></p>`

**Comment by pauleveritt** [comment]
Note: Added an important "last point" at the end.

**Comment by Finistere** [comment]
I think `@lazy` does what you're searching for:

```python
from antidote import const, inject, lazy, world


class Conf:
    PUNCTUATION = const('!')


@lazy
def Greeting(name: str, punctuation: str = inject[Conf.PUNCTUATION]) -> str:
    return f"<p>Hello {name}{punctuation}</p>"


assert world[Greeting(name='Antidote')] == "<p>Hello Antidote!</p>"
assert world[Greeting(name='someone', punctuation='???')] == "<p>Hello someone???</p>"
```
